### PR TITLE
Fix/handle keychain errors

### PR DESCRIPTION
**Pull request by achocron**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This PR fixes an issue where, in certain cases, the app does not have access to the update token stored in keychain. This happens when the phone is locked with a passcode but the flow is executing in the background. We've seen it a lot when there are silent pushes.

The side effects of this is that when in app updates fails to retrieve the token but tries to get release details anyways, a 403 is received because the authentication fails without the token. Then, all state is cleared and the user must sign in to app center on relaunch.

This PR fixes that by
1. enabling a mode where the update token *can* be accessed in locked mode (simple keychain flag)
2. there could still be an edge case where it is not accessed so when we see the access error, don't clear state

[AB#72031](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/72031)

### CODE REVIEW

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m [hunk 1]
```diff
@@ -239,7 +239,12 @@ - (void)startUpdate {
   NSString *releaseHash = MSPackageHash();
   if (releaseHash) {
     [self changeDistributionGroupIdAfterAppUpdateIfNeeded:releaseHash];
-    NSString *updateToken = [MSKeychainUtil stringForKey:kMSUpdateTokenKey];
+    OSStatus statusCode;
+    NSString *updateToken = [MSKeychainUtil stringForKey:kMSUpdateTokenKey statusCode:&statusCode];
+    if (statusCode == errSecInteractionNotAllowed) {
```
Comment (jaelim-ms) [inline AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: Shouldn't you log errors for all failed status codes?

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m [hunk 2]
```diff
@@ -439,7 +444,6 @@ - (void)checkLatestRelease:(NSString *)updateToken distributionGroupId:(NSString
                     // If the response payload is MSErrorDetails, consider it as a recoverable error.
                     if (!details || ![kMSErrorCodeNoReleasesForUser isEqualToString:details.code]) {
                       [MSKeychainUtil deleteStringForKey:kMSUpdateTokenKey];
-                      [MS_USER_DEFAULTS removeObjectForKey:kMSSDKHasLaunchedWithDistribute];
```
Comment (jaelim-ms) [inline AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: Any reason of removing this line?

File: AppCenter/AppCenterTests/Util/MSMockKeychainUtil.m [hunk 3]
```diff
@@ -68,14 +71,31 @@ + (NSString *_Nullable)deleteStringForKey:(NSString *)key withServiceName:(NSStr
...
+  if (statusCode) {
+    *statusCode = placeholderStatus;
+  }
+  if (placeholderStatus != noErr) {
+    return nil;
+  }
   return stringsDictionary[serviceName][key];
 }
 
++ (void)mockStatusCode:(OSStatus)statusCode forKey:(NSString *)key {
```
Comment (jaelim-ms) [inline AppCenter/AppCenterTests/Util/MSMockKeychainUtil.m]: Why do you need this one? Can't you just remove all objects in `stopMocking ` method and use it?

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 4]
```diff
@@ -2324,6 +2324,56 @@ - (void)testShouldChangeDistributionGroupIdIfStoredIdDoesntMatchDownloadedId {
...
+  OCMReject([distributeMock requestInstallInformationWith:OCMOCK_ANY]);
+  OCMReject([distributeMock checkLatestRelease:OCMOCK_ANY distributionGroupId:OCMOCK_ANY releaseHash:OCMOCK_ANY]);
+
+  // When
+  [distributeMock startUpdate];
+
+  // Stop mocking
+  [distributeMock stopMocking];
+  [utilityMock stopMocking];
+  [MSMockKeychainUtil mockStatusCode:noErr forKey:kMSUpdateTokenKey];
```
Comment (jaelim-ms) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: If you use keychain mock, you must call `stopMocking` to clean it up.

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m (hunk 1 above)
Comment (achocron) [reply AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: Keychain util will log.

File: AppCenter/AppCenterTests/Util/MSMockKeychainUtil.m (hunk 3 above)
Comment (achocron) [reply AppCenter/AppCenterTests/Util/MSMockKeychainUtil.m]: We specifically need to mock certain status codes - this isn't only used for cleanup

Comment (achocron) [reply AppCenter/AppCenterTests/Util/MSMockKeychainUtil.m]: but we can do that. still need to keep the method though.

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m (hunk 2 above)
Comment (achocron) [reply AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: The user will be re-prompted to sign in after an unrecoverable error. This was causing issues but i am not sure if we need that line to be removed now that we fixed the issue at another level.

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m (hunk 1 above)
Comment (achocron) [reply AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: The logic here is more about returning than logging.

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m (hunk 2 above)
Comment (achocron) [reply AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: I can re-add it. It probably isn't needed anymore.

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m (hunk 4 above)
Comment (achocron) [reply AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: It is called in tearDown

**Pull request merged**